src/contracts/core/PaymentCoordinator.sol

```diff
@@ -0,0 +1,502 @@
+        RangePayment[] calldata rangePayments
+    ) external onlyWhenNotPaused(PAUSED_PAY_FOR_RANGE) nonReentrant {
+        for (uint256 i = 0; i < rangePayments.length; i++) {
+            RangePayment calldata rangePayment = rangePayments[i];
+            bytes32 rangePaymentHash = keccak256(abi.encode(msg.sender, paymentNonce, rangePayment));
+            require(
+                !isRangePaymentHash[msg.sender][rangePaymentHash],
+                "PaymentCoordinator._payForRange: range payment hash already submitted"
+            );
+            isRangePaymentHash[msg.sender][rangePaymentHash] = true;
+
+            _payForRange(rangePayment);
+            emit RangePaymentCreated(msg.sender, paymentNonce, rangePaymentHash, rangePayment);
+            paymentNonce++;
+        }
+    }
+
+    /**
+     * @notice similar to `payForRange` except the payment is split amongst *all* stakers
+     * rather than just those delegated to operators who are registered to a single avs and is
+     * a permissioned call based on isPayAllForRangeSubmitter mapping.
+     * @param rangePayments The range payments being created
+     */
+    function payAllForRange(
+        RangePayment[] calldata rangePayments
+    ) external onlyWhenNotPaused(PAUSED_PAY_ALL_FOR_RANGE) onlyPayAllForRangeSubmitter nonReentrant {
+        for (uint256 i = 0; i < rangePayments.length; i++) {
+            RangePayment calldata rangePayment = rangePayments[i];
+            bytes32 rangePaymentForAllHash = keccak256(abi.encode(msg.sender, paymentNonce, rangePayment));
+            require(
+                !isRangePaymentForAllHash[msg.sender][rangePaymentForAllHash],
+                "PaymentCoordinator._payForRange: range payment hash already submitted"
+            );
+            isRangePaymentForAllHash[msg.sender][rangePaymentForAllHash] = true;
+
+            _payForRange(rangePayment);
+            emit RangePaymentForAllCreated(msg.sender, paymentNonce, rangePaymentForAllHash, rangePayment);
+            paymentNonce++;
+        }
+    }
+
+    /**
+     * @notice Claim payments against a given root (read from distributionRoots[claim.rootIndex]).
+     * Earnings are cumulative so earners don't have to claim against all distribution roots they have earnings for,
+     * they can simply claim against the latest root and the contract will calculate the difference between
+     * their cumulativeEarnings and cumulativeClaimed. This difference is then transferred to claimerFor[claim.earner]
+     * @param claim The PaymentMerkleClaim to be processed.
+     * Contains the root index, earner, payment leaves, and required proofs
+     * @dev only callable by the valid claimer, that is
+     * if claimerFor[claim.earner] is address(0) then only the earner can claim, otherwise only
+     * claimerFor[claim.earner] can claim the payments.
+     */
+    function processClaim(PaymentMerkleClaim calldata claim) external onlyWhenNotPaused(PAUSED_CLAIM_PAYMENTS) {
```

Comment:
Would prefer to not immediately call an internal function here - unless there's a need I'm missing? This seems to be the only place we call `_processClaim`

Reply:
[6f3d65e](https://github.com/Layr-Labs/eigenlayer-contracts/pull/514/commits/6f3d65e2dc454dcc0a3a57451e003455a69fa0fb)